internal/server/email/smtp.go

```diff
@@ -0,0 +1,222 @@
+package email
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"net/smtp"
+	"strings"
+	"sync"
+	"time"
+)
+
+type Client struct {
+	*tls.Conn
+	reader *bufio.Reader
+	sync.Mutex
+	skipTLSVerify bool
+}
+
+type Message struct {
+	FromName    string
+	FromAddress string
+	ToName      string
+	ToAddress   string
+	Subject     string
+	PlainBody   []byte
+	HTMLBody    []byte
+}
+
+var client = &Client{}
+
+func (c *Client) connect() (err error) {
+	if c.Conn != nil {
+		return nil
+	}
+
+	if SendgridAPIKey == "" {
+		return fmt.Errorf("Sendgrid API key is not set")
+	}
+
+	conn, err := tls.Dial("tcp", SMTPServer, &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
+		// nolint:gosec
+		InsecureSkipVerify: c.skipTLSVerify, // only used by tests
+	})
+	if err != nil {
+		return err
+	}
+
+	err = conn.SetDeadline(time.Now().Add(30 * time.Second))
+	if err != nil {
+		return fmt.Errorf("set deadline: %w", err)
+	}
+	c.Conn = conn
+
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	auth := smtp.PlainAuth("", "apikey", SendgridAPIKey, SMTPServer)
+
+	client, err := smtp.NewClient(conn, SMTPServer)
+	if err != nil {
+		return err
+	}
+
+	if err = client.Auth(auth); err != nil {
+		return err
+	}
+
+	c.reader = bufio.NewReader(c.Conn)
+
+	return nil
+}
+
+func base64encode(s string) string {
+	return base64.StdEncoding.EncodeToString([]byte(s))
+}
+
+// readln reads a line of text ending in \n. if expected is empty it returns the result.
+// if expected is not empty, the read string must match or start with the expected string
+func (c *Client) readln(expected string) (result string, err error) {
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	if err := c.SetDeadline(time.Now().Add(30 * time.Second)); err != nil {
+		return "", fmt.Errorf("set deadline: %w", err)
+	}
+
+	s, err := c.reader.ReadString('\n')
+	if err != nil {
+		return "", err
+	}
+
+	s = strings.TrimRight(s, "\r\n")
+
+	if expected != "" && !strings.HasPrefix(s, expected) {
+		return s, fmt.Errorf("Unexpected value read: %q while waiting for %q", s, expected)
+	}
+
+	return s, nil
+}
+
+func (c *Client) writeln(s string) (err error) {
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	s += "\r\n"
+	if err = c.SetWriteDeadline(time.Now().Add(30 * time.Second)); err != nil {
+		return fmt.Errorf("set write deadline: %w", err)
+	}
+	written, err := c.Write([]byte(s))
+	if err != nil {
+		return fmt.Errorf("write auth login: %w", err)
+	}
+	if written != len(s) {
+		return fmt.Errorf("partial write %d bytes of %d", written, len(s))
+	}
+
+	return nil
+}
+
+// SendSMTP sends an email message
+func SendSMTP(msg Message) error {
+	client.Lock()
+	defer client.Unlock()
+
+	if err := client.connect(); err != nil {
+		return err
+	}
+
+	if err := client.writeln(fmt.Sprintf("MAIL FROM: %s", msg.FromAddress)); err != nil {
```

Comment:
Why not use `smtp.SendMail()` instead?

Reply:
it was just so simple. :D I can change it though.

Reply:
So one problem I'm running into with smtp.SendMail is that it assumes :smtp port (25) and then tries to upgrade to tls if the server supports it. This is different from what I'm doing; I'm connecting to a TLS-encrypted SMTP port and failing if it doesn't connect. Switching to this other code means there's a potential silent downgrade to non-TLS if the destination server doesn't support TLS